ndmg/utils/bids_utils.py

```diff
@@ -0,0 +1,175 @@
+
+from bids.grabbids import BIDSLayout
+import re
+from itertools import product
+
+
+def sweep_directory(bdir, subj=None, sesh=None, task=None, run=None, modality='dwi'):
+    """
+    Given a BIDs formatted directory, crawls the BIDs dir and prepares the
+    necessary inputs for the NDMG pipeline. Uses regexes to check matches for
+    BIDs compliance.
+    """
+    if modality == 'dwi':
+        dwis = []
+        bvals = []
+        bvecs = []
+    elif modality == 'func':
+        funcs = []
+    anats = []
+    layout = BIDSLayout(bdir)  # initialize BIDs tree on bdir
+    # get all files matching the specific modality we are using
+    if subj is None:
+        subjs = layout.get_subjects()  # list of all the subjects
+    else:
+        subjs = as_list(subj)  # make it a list so we can iterate
+    for sub in subjs:
+        if not sesh:
+            seshs = layout.get_sessions(subject=sub)
+            seshs += [None]  # in case there are non-session level inputs
+        else:
+            seshs = as_list(sesh)  # make a list so we can iterate
+
+        if not task:
+            tasks = layout.get_tasks(subject=sub)
+            tasks += [None]
+        else:
+            tasks = as_list(task)
+
+        if not run:
+            runs = layout.get_runs(subject=sub)
+            runs += [None]
+        else:
+            runs = as_list(run)
+
+        # all the combinations of sessions and tasks that are possible
+        for (ses, tas, ru) in product(seshs, tasks, runs):
+            # the attributes for our modality img
+            mod_attributes = [sub, ses, tas, ru]
+            # the keys for our modality img
+            mod_keys = ['subject', 'session', 'task', 'run']
+            # our query we will use for each modality img
+            mod_query = {'modality': modality}
+            if modality == 'dwi':
+                type_img = 'dwi'  # use the dwi image
+            elif modality == 'func':
+                type_img = 'bold'  # use the bold image
+            mod_query['type'] = type_img
+
+            for attr, key in zip(mod_attributes, mod_keys):
+                if attr:
+                    mod_query[key] = attr
+
+            anat_attributes = [sub, ses]  # the attributes for our anat img
+            anat_keys = ['subject', 'session']  # the keys for our modality img
+            # our query for the anatomical image
+            anat_query = {'modality': 'anat', 'type': 'T1w',
+                          'extensions': 'nii.gz|nii'}
+            for attr, key in zip(anat_attributes, anat_keys):
+                if attr:
+                    anat_query[key] = attr
+            # make a query to fine the desired files from the BIDSLayout
+            anat = layout.get(**anat_query)
+            if modality == 'dwi':
+                dwi = layout.get(**merge_dicts(mod_query,
+                                               {'extensions': 'nii.gz|nii'}))
+                bval = layout.get(**merge_dicts(mod_query,
```

Comment:
I don't believe this will work in the case where a bval/bvec file exists only in the upstream directory, not side-by-side with the niftii. Also, I don't really understand why you're using this `merge_dicts` function you made - I think it would be good to revisit the `pybids` repository and explore how *they* recommend using it, and using more of their functions. For instance, `get_bval` when provided with a dwi volume file will get the path for the correct associated file. Similarly for bvecs, etc....

Reply:
merge_dicts allows you to query for any combination of subject, session, task, and run level specification, since you need to pass variable dictionaries to pybids to get an appropriate query together with layout.get(...) depending on the specifications you have available

Reply:
This still doesn't address that your code doesn't work for valid diffusion configurations.

The script I had sent you this morning does have code using pybids that handles this better, and the BIDS app specification does not enable granularity finer than the `session` level anyways.

I.e., you need to just do:
 - `get_subjects()` and `get_sessions()`, then iterate over the two, doing 
   - `get(subject=###, session=###, extensions="nii|nii.gz")` and `get_bval()`/`get_bvec()` or `get_metadata()/get_events()/get_task()` on the corresponding DWI or BOLD image, respectively.